tests/pkg_paho-mqtt/main.c

```diff
@@ -0,0 +1,306 @@
+        NetworkDisconnect(&network);
+    }
+
+    int port = DEFAULT_MQTT_PORT;
+    if (argc > 2) {
+        port = atoi(argv[2]);
+    }
+
+    MQTTPacket_connectData data = MQTTPacket_connectData_initializer;
+    data.MQTTVersion = MQTT_VERSION_v311;
+
+    data.clientID.cstring = DEFAULT_MQTT_CLIENT_ID;
+    if (argc > 3) {
+        data.clientID.cstring = argv[3];
+    }
+
+    data.username.cstring = DEFAULT_MQTT_USER;
+    if (argc > 4) {
+        data.username.cstring = argv[4];
+    }
+
+    data.password.cstring = DEFAULT_MQTT_PWD;
+    if (argc > 5) {
+        data.password.cstring = argv[5];
+    }
+
+    data.keepAliveInterval = DEFAULT_KEEPALIVE_SEC;
+    if (argc > 6) {
+        data.keepAliveInterval = atoi(argv[6]);
+    }
+
+    data.cleansession = IS_CLEAN_SESSION;
+    data.willFlag = 0;
+
+    ret = NetworkConnect(&network, remote_ip, port);
+    if (ret < 0) {
+        printf("error: unable to connect\n");
+        return ret;
+    }
+
+    ret = MQTTConnect(&client, &data);
+    if (ret < 0) {
+        printf("error: unable to connect to broker (%d)\n", ret);
+        _discon(0, NULL);
+        return ret;
+    }
+    else {
+        printf("success: connected to broker");
+    }
+
+    return (ret > 0) ? 0 : 1;
+}
+
+static int _discon(int argc, char **argv)
+{
+    (void)argc;
+    (void)argv;
+
+    topic_cnt = 0;
+    int res = MQTTDisconnect(&client);
+    if (res < 0) {
+        printf("error: unable to disconnect\n");
+    }
+    else {
+        printf("success: disconnected from broker\n");
+    }
+
+    NetworkDisconnect(&network);
+    return res;
+}
+
+static int _pub(int argc, char **argv)
+{
+    enum QoS qos = QOS0;
+
+    if (argc < 3) {
+        printf("usage: %s <topic name> <string msg> [QoS level]\n",
+               argv[0]);
+        return 1;
+    }
+    if (argc == 4) {
+        qos = get_qos(argv[3]);
+    }
+    MQTTMessage message;
+    message.qos = qos;
+    message.retained = IS_RETAINED_MSG;
+    message.payload = argv[2];
+    message.payloadlen = strlen(message.payload);
+
+    int rc;
+    if ((rc = MQTTPublish(&client, argv[1], &message)) < 0) {
+        printf("error: unable to publish (%d)\n", rc);
+    }
+    else {
+        printf("success: published message '%s' to topic '%s' with QoS %d\n",
+               (char *)message.payload, argv[1], (int)message.qos);
+    }
+
+    return rc;
+}
+
+static int _sub(int argc, char **argv)
+{
+    enum QoS qos = QOS0;
+
+    if (argc < 2) {
+        printf("usage: %s <topic name> [QoS level]\n", argv[0]);
+        return 1;
+    }
+
+    if (argc >= 3) {
+        qos = get_qos(argv[2]);
+    }
+
+    if (topic_cnt > MAX_TOPICS) {
+        printf("error: already subscribed to max %d topics\n", topic_cnt);
+        return -1;
+    }
+
+    if (strlen(argv[1]) > MAX_LEN_TOPIC) {
+        printf("error: not subscribing, topic too long %s\n", argv[1]);
+        return -1;
+    }
+    strncpy(_topic_to_subscribe[topic_cnt], argv[1], strlen(argv[1]));
```

Comment:
This makes little sense. It should be at least
```suggestion
    strncpy(_topic_to_subscribe[topic_cnt], argv[1], MAX_LEN_TOPIC);
```

But then again, if `strlen()` is called above and the length is checked anyway, one could use the more efficient `memcpy()` instead.

It also looks to me like the topics should be zero-terminated, so it should be `strlen(argv[1)) + 1 > MAX_LEN_TOPIC`, not `strlen(argv[1)) > MAX_LEN_TOPIC`

Reply:
Agreed. I made both changes (but kept the `strncpy`).

Same question as above: This code line originates from the `examples/paho-mqtt/main.c` which I based the test's `main.c` on. Should I also make this change there or should that be a second (very small) PR?